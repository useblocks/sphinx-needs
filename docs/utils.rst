.. _utils:

Utils
=====

.. toctree::
   :maxdepth: 2

   performance/index
   performance/debug
