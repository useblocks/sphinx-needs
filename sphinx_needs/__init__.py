from sphinx_needs.needs import __version__, setup  # noqa: F401
