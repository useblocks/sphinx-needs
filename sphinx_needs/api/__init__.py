# flake8: noqa

from .configuration import (
    add_dynamic_function,
    add_extra_option,
    add_need_type,
    get_need_types,
)
from .need import add_external_need, add_need, del_need, make_hashed_id
