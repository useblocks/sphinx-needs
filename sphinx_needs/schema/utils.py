from sphinx_needs.schema.config import NeedFieldsSchemaWithVersionType


def get_properties_from_schema(
    schema: NeedFieldsSchemaWithVersionType,
) -> set[str]:
    """
    Extract a list of property names from a given JSON schema.

    It searches both the top-level "properties" and any nested properties in "allOf" schemas.
    """
    properties: set[str] = set()

    # Extract properties in the main "properties" field
    if "properties" in schema and isinstance(schema["properties"], dict):
        properties.update(schema["properties"].keys())

    # If there is an "allOf" key with additional schemas, extract their properties as well
    if "allOf" in schema and isinstance(schema["allOf"], list):
        for subschema in schema["allOf"]:
            assert "$ref" not in subschema, "$ref have already been resolved"
            if "properties" in subschema and isinstance(subschema["properties"], dict):
                properties.update(subschema["properties"].keys())

    return properties
