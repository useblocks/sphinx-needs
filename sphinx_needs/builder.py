import os
from typing import Iterable, Optional, Set

from docutils import nodes
from sphinx import version_info
from sphinx.application import Sphinx
from sphinx.builders import Builder

from sphinx_needs.logging import get_logger
from sphinx_needs.needsfile import NeedsList
from sphinx_needs.utils import unwrap

log = get_logger(__name__)


class NeedsBuilder(Builder):
    name = "needs"
    format = "json"
    file_suffix = ".txt"
    links_suffix = None

    def write_doc(self, docname: str, doctree: nodes.document) -> None:
        pass

    def finish(self) -> None:
        env = unwrap(self.env)
        needs = env.needs_all_needs.values()  # We need a list of needs for later filter checks
        filters = env.needs_all_filters
        config = env.config
        version = getattr(config, "version", "unset")
        needs_list = NeedsList(config, self.outdir, self.srcdir)

        if config.needs_file:
            needs_file = config.needs_file
            needs_list.load_json(needs_file)
        else:
            # check if needs.json file exists in conf.py directory
            needs_json = os.path.join(self.srcdir, "needs.json")
            if os.path.exists(needs_json):
                log.info("needs.json found, but will not be used because needs_file not configured.")

        # Clean needs_list from already stored needs of the current version.
        # This is needed as needs could have been removed from documentation and if this is the case,
        # removed needs would stay in needs_list, if list gets not cleaned.
        needs_list.wipe_version(version)
        #
        from sphinx_needs.filter_common import filter_needs

        filter_string = self.app.config.needs_builder_filter
        filtered_needs = filter_needs(self.app, needs, filter_string)

        for need in filtered_needs:
            needs_list.add_need(version, need)

        for need_filter in filters.values():
            if need_filter["export_id"]:
                needs_list.add_filter(version, need_filter)

        try:
            needs_list.write_json()
        except Exception as e:
            log.error(f"Error during writing json file: {e}")
        else:
            log.info("Needs successfully exported")

    def get_outdated_docs(self) -> Iterable[str]:
        return []

    def prepare_writing(self, _docnames: Set[str]) -> None:
        pass

    def write_doc_serialized(self, _docname: str, _doctree: nodes.document) -> None:
        pass

    def cleanup(self) -> None:
        pass

    def get_target_uri(self, _docname: str, _typ: Optional[str] = None) -> str:
        return ""


def build_needs_json(app: Sphinx, _exception: Exception) -> None:

    env = unwrap(app.env)

    if not env.config.needs_build_json:
        return

    # Do not create an additional needs.json, if builder is already "needs".
    if isinstance(app.builder, NeedsBuilder):
        return

    try:
        needs_builder = NeedsBuilder(app, env)
    except TypeError:
        needs_builder = NeedsBuilder(app)
        needs_builder.set_environment(env)

    needs_builder.finish()


class NeedumlsBuilder(Builder):
    name = "needumls"

    def write_doc(self, docname: str, doctree: nodes.document) -> None:
        pass

    def finish(self) -> None:
        env = unwrap(self.env)
        needumls = env.needs_all_needumls.values()

        for needuml in needumls:
            if needuml["save"]:
                puml_content = needuml["content_calculated"]
                # check if given save path dir exists
                save_path = os.path.join(self.outdir, needuml["save"])
                save_dir = os.path.dirname(save_path)
                if not os.path.exists(save_dir):
                    os.makedirs(save_dir, exist_ok=True)

                log.info(f"Storing needuml data to file {save_path}.")
                with open(save_path, "w") as f:
                    f.write(puml_content)

    def get_outdated_docs(self) -> Iterable[str]:
        return []

    def prepare_writing(self, _docnames: Set[str]) -> None:
        pass

    def write_doc_serialized(self, _docname: str, _doctree: nodes.document) -> None:
        pass

    def cleanup(self) -> None:
        pass

    def get_target_uri(self, _docname: str, _typ: Optional[str] = None) -> str:
        return ""


def build_needumls_pumls(app: Sphinx, _exception: Exception) -> None:
    env = unwrap(app.env)

    if not env.config.needs_build_needumls:
        return

    # Do not create additional files for saved plantuml content, if builder is already "needumls".
    if isinstance(app.builder, NeedumlsBuilder):
        return

    # if other builder like html used together with config: needs_build_needumls
    if version_info[0] >= 5:
        needs_builder = NeedumlsBuilder(app, env)
        needs_builder.outdir = os.path.join(needs_builder.outdir, env.config.needs_build_needumls)
    else:
        needs_builder = NeedumlsBuilder(app)
        needs_builder.outdir = os.path.join(needs_builder.outdir, env.config.needs_build_needumls)
        needs_builder.set_environment(env)

    needs_builder.finish()
