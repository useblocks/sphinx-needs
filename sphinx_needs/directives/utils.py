import re
from typing import Any, Dict, List, Tuple

from docutils import nodes
from sphinx.environment import BuildEnvironment

from sphinx_needs.defaults import TITLE_REGEX


def no_needs_found_paragraph() -> nodes.paragraph:
    nothing_found = "No needs passed the filters"
    para = nodes.paragraph()
    nothing_found_node = nodes.Text(nothing_found)
    para += nothing_found_node
    return para


def used_filter_paragraph(current_needfilter) -> nodes.paragraph:
    para = nodes.paragraph()
    filter_text = "Used filter:"
    filter_text += (
        " status(%s)" % " OR ".join(current_needfilter["status"]) if len(current_needfilter["status"]) > 0 else ""
    )
    if len(current_needfilter["status"]) > 0 and len(current_needfilter["tags"]) > 0:
        filter_text += " AND "
    filter_text += " tags(%s)" % " OR ".join(current_needfilter["tags"]) if len(current_needfilter["tags"]) > 0 else ""
    if (len(current_needfilter["status"]) > 0 or len(current_needfilter["tags"]) > 0) and len(
        current_needfilter["types"]
    ) > 0:
        filter_text += " AND "
    filter_text += (
        " types(%s)" % " OR ".join(current_needfilter["types"]) if len(current_needfilter["types"]) > 0 else ""
    )

    filter_node = nodes.emphasis(filter_text, filter_text)
    para += filter_node
    return para


def get_link_type_option(name: str, env: BuildEnvironment, node, default: str = "") -> List[str]:
    link_types = [x.strip() for x in re.split(";|,", node.options.get(name, default))]
    conf_link_types = env.config.needs_extra_links
    conf_link_types_name = [x["option"] for x in conf_link_types]

    final_link_types = []
    for link_type in link_types:
        if link_type is None or link_type == "":
            continue
        if link_type not in conf_link_types_name:
            raise SphinxNeedsLinkTypeException(link_type + "does not exist in configuration option needs_extra_links")

        final_link_types.append(link_type)
    return final_link_types


def get_title(option_string: str) -> Tuple[str, str]:
    """
    Returns a tuple of uppercase option and calculated title of given option string.

    :param option_string:
    :return: string
    """
    if option_string.upper() == "ID":
        return "ID", "ID"
    match = re.search(TITLE_REGEX, option_string)
    if not match:
        return option_string.upper(), option_string.title().replace("_", " ")

    option_name = match.group(1)
    title = match.group(2)

    return option_name.upper(), title


def get_option_list(options, name: str) -> List[str]:
    """
    Gets and creates a list of a given directive option value in a safe way
    :param options: List of options
    :param name: Name of the option
    :return: List with strings
    """
    values = str(options.get(name, ""))
    values_list = []
    if isinstance(values, str):
        values_list = [value.strip() for value in re.split("[;,]", values)]

    return values_list


def analyse_needs_metrics(env: BuildEnvironment) -> Dict[str, Any]:
    """
    Function to generate metrics about need objects.

    :param env: Sphinx build environment
    :return: Dictionary consisting of needs metrics.
    """
    needs: Dict = env.needs_all_needs
    metric_data = {"needs_amount": len(needs)}
    needs_types = {i["directive"]: 0 for i in env.app.config.needs_types}

    for i in needs.values():
        if i["type"] in needs_types:
            needs_types[i["type"]] += 1

    metric_data["needs_types"] = {i[0]: i[1] for i in sorted(needs_types.items(), key=lambda x: x[0])}
    return metric_data


class SphinxNeedsLinkTypeException(BaseException):
    """Raised if problems with link types happen"""
