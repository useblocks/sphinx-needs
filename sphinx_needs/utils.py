import cProfile
import importlib
import operator
import os
import re
from functools import reduce, wraps
from re import Pattern
from typing import Any, Dict, List, Optional, TypeVar, Union
from urllib.parse import urlparse

from docutils import nodes
from jinja2 import BaseLoader, Environment, Template
from sphinx.application import BuildEnvironment, Sphinx

from sphinx_needs.defaults import NEEDS_PROFILING
from sphinx_needs.logging import get_logger

logger = get_logger(__name__)

NEEDS_FUNCTIONS = {}

# List of internal need option names. They should not be used by or presented to user.
INTERNALS = [
    "docname",
    "doctype",
    "lineno",
    "target_node",
    "refid",
    "content",
    "pre_content",
    "post_content",
    "collapse",
    "parts",
    "id_parent",
    "id_complete",
    "title",
    "full_title",
    "is_part",
    "is_need",
    "type_prefix",
    "type_color",
    "type_style",
    "type",
    "type_name",
    "id",
    "hide",
    "hide_status",
    "hide_tags",
    "sections",
    "section_name",
    "content_node",
    # "parent_needs",
    "parent_need",
    # "child_needs",
    "is_external",
    "external_css",
    "is_modified",
    "modifications",
    "constraints",
    "constraints_passed",
    "constraints_results",
    "arch",
]

MONTH_NAMES = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
]


def row_col_maker(
    app: Sphinx,
    fromdocname: str,
    all_needs,
    need_info,
    need_key,
    make_ref: bool = False,
    ref_lookup: bool = False,
    prefix: str = "",
):
    """
    Creates and returns a column.

    :param app: current sphinx app
    :param fromdocname: current document
    :param all_needs: Dictionary of all need objects
    :param need_info: need_info object, which stores all related need data
    :param need_key: The key to access the needed data from need_info
    :param make_ref: If true, creates a reference for the given data in need_key
    :param ref_lookup: If true, it uses the data to lookup for a related need and uses its data to create the reference
    :param prefix: string, which is used as prefix for the text output
    :return: column object (nodes.entry)
    """
    builder = unwrap(app.builder)
    env = unwrap(builder.env)

    row_col = nodes.entry(classes=["needs_" + need_key])
    para_col = nodes.paragraph()

    needs_string_links_option: List[str] = []
    for v in app.config.needs_string_links.values():
        needs_string_links_option.extend(v["options"])

    if need_key in need_info and need_info[need_key] is not None:
        if isinstance(need_info[need_key], (list, set)):
            data = need_info[need_key]
        elif isinstance(need_info[need_key], str) and need_key in needs_string_links_option:
            data = re.split(r",|;", need_info[need_key])
            data = [i.strip() for i in data if len(i) != 0]
        else:
            data = [need_info[need_key]]

        for index, datum in enumerate(data):
            link_id = datum
            link_part = None

            link_list = []
            for link_type in env.config.needs_extra_links:
                link_list.append(link_type["option"])
                link_list.append(link_type["option"] + "_back")

            # For needs_string_links
            link_string_list = {}
            for link_name, link_conf in app.config.needs_string_links.items():
                link_string_list[link_name] = {
                    "url_template": Environment(loader=BaseLoader, autoescape=True).from_string(link_conf["link_url"]),
                    "name_template": Environment(loader=BaseLoader, autoescape=True).from_string(
                        link_conf["link_name"]
                    ),
                    "regex_compiled": re.compile(link_conf["regex"]),
                    "options": link_conf["options"],
                    "name": link_name,
                }

            matching_link_confs = []
            for link_conf in link_string_list.values():
                if need_key in link_conf["options"] and len(datum) != 0:
                    matching_link_confs.append(link_conf)

            if need_key in link_list and "." in datum:
                link_id = datum.split(".")[0]
                link_part = datum.split(".")[1]

            datum_text = prefix + str(datum)
            text_col = nodes.Text(datum_text)
            if make_ref or ref_lookup:
                try:
                    if need_info["is_external"]:
                        ref_col = nodes.reference("", "")
                    else:
                        # Mark references as "internal" so that if the rinohtype builder is being used it produces an
                        # internal reference within the generated PDF instead of an external link. This replicates the
                        # behaviour of references created with the sphinx utility `make_refnode`.
                        ref_col = nodes.reference("", "", internal=True)

                    if make_ref:
                        if need_info["is_external"]:
                            ref_col["refuri"] = check_and_calc_base_url_rel_path(need_info["external_url"], fromdocname)
                            ref_col["classes"].append(need_info["external_css"])
                            row_col["classes"].append(need_info["external_css"])
                        else:
                            ref_col["refuri"] = builder.get_relative_uri(fromdocname, need_info["docname"])
                            ref_col["refuri"] += "#" + datum
                    elif ref_lookup:
                        temp_need = all_needs[link_id]
                        if temp_need["is_external"]:
                            ref_col["refuri"] = check_and_calc_base_url_rel_path(temp_need["external_url"], fromdocname)
                            ref_col["classes"].append(temp_need["external_css"])
                            row_col["classes"].append(temp_need["external_css"])
                        else:
                            ref_col["refuri"] = builder.get_relative_uri(fromdocname, temp_need["docname"])
                            ref_col["refuri"] += "#" + temp_need["id"]
                            if link_part:
                                ref_col["refuri"] += "." + link_part

                except KeyError:
                    para_col += text_col
                else:
                    ref_col.append(text_col)
                    para_col += ref_col
            elif matching_link_confs:
                para_col += match_string_link(
                    datum_text, datum, need_key, matching_link_confs, render_context=app.config.needs_render_context
                )
            else:
                para_col += text_col

            if index + 1 < len(data):
                para_col += nodes.emphasis("; ", "; ")

    row_col += para_col

    return row_col


def rstjinja(app: Sphinx, docname: str, source: List[str]) -> None:
    """
    Render our pages as a jinja template for fancy templating goodness.
    """
    builder = unwrap(app.builder)
    # Make sure we're outputting HTML
    if builder.format != "html":
        return
    src = source[0]
    rendered = builder.templates.render_string(src, app.config.html_context, **app.config.needs_render_context)
    source[0] = rendered


def import_prefix_link_edit(needs: Dict[str, Any], id_prefix: str, needs_extra_links: List[Dict[str, Any]]) -> None:
    """
    Changes existing links to support given prefix.
    Only link-ids get touched, which are part of ``needs`` (so are linking them).
    Other links do not get the prefix, as they are treated as "external" links.

    :param needs: Dict of all needs
    :param id_prefix: Prefix as string
    :param needs_extra_links: config var of all supported extra links. Normally coming from env.config.needs_extra_links
    :return:
    """
    needs_ids = needs.keys()

    for need in needs.values():
        for id in needs_ids:
            # Manipulate links in all link types
            for extra_link in needs_extra_links:
                if extra_link["option"] in need and id in need[extra_link["option"]]:
                    for n, link in enumerate(need[extra_link["option"]]):
                        if id == link:
                            need[extra_link["option"]][n] = f"{id_prefix}{id}"
            # Manipulate descriptions
            # ToDo: Use regex for better matches.
            need["description"] = need["description"].replace(id, "".join([id_prefix, id]))


def profile(keyword: str):
    """
    Activate profiling for a specific function.

    Activation only happens, if given keyword is part of ``needs_profiling``.
    """

    def inner(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            with cProfile.Profile() as pr:
                result = func(*args, **kwargs)

            profile_folder = os.path.join(os.getcwd(), "profile")
            profile_file = os.path.join(profile_folder, f"{keyword}.prof")
            if not os.path.exists(profile_file):
                os.makedirs(profile_folder, exist_ok=True)
            pr.dump_stats(profile_file)
            return result

        if keyword in NEEDS_PROFILING:
            return wrapper
        return func

    return inner


def check_and_calc_base_url_rel_path(external_url: str, fromdocname: str) -> str:
    """
    Check given base_url from needs_external_needs and calculate relative path if base_url is relative path.

    :param external_url: Caculated external_url from base_url in needs_external_needs
    :param fromdocname: document name
    :return: calculated external_url
    """
    ref_uri = external_url
    # check if given base_url is url or relative path
    parsed_url = urlparse(external_url)
    # get path sep considering plattform dependency, '\' for Windows, '/' fro Unix
    curr_path_sep = os.path.sep
    # check / or \ to determine the relative path to conf.py directory
    if not parsed_url.scheme and not os.path.isabs(external_url) and curr_path_sep in fromdocname:
        sub_level = len(fromdocname.split(curr_path_sep)) - 1
        ref_uri = os.path.join(sub_level * (".." + curr_path_sep), external_url)

    return ref_uri


def check_and_get_external_filter_func(current_needlist):
    """Check and import filter function from external python file."""
    # Check if external filter code is defined
    filter_func = None
    filter_args = []

    filter_func_ref = current_needlist.get("filter_func", None)

    if filter_func_ref:
        try:
            filter_module, filter_function = filter_func_ref.rsplit(".")
        except ValueError:
            logger.warning(f'Filter function not valid "{filter_func_ref}". Example: my_module:my_func')
            return []  # No needs were found because of invalid filter function

        result = re.search(r"^(\w+)(?:\((.*)\))*$", filter_function)
        filter_function = result.group(1)
        filter_args = result.group(2) or []

        try:
            final_module = importlib.import_module(filter_module)
            filter_func = getattr(final_module, filter_function)
        except Exception:
            logger.warning(f"Could not import filter function: {filter_func_ref}")
            return []

    return filter_func, filter_args


def jinja_parse(context: Dict, jinja_string: str) -> str:
    """
    Function to parse mapping options set to a string containing jinja template format.

    :param context: Data to be used as context in rendering jinja template
    :type context: dict
    :param jinja_string: A jinja template string
    :type jinja_string: str
    :return: A rendered jinja template as string
    :rtype: str

    """
    try:
        content_template = Template(jinja_string, autoescape=True)
    except Exception as e:
        raise ReferenceError(f'There was an error in the jinja statement: "{jinja_string}". ' f"Error Msg: {e}")

    content = content_template.render(**context)
    return content


def dict_get(root, items, default=None) -> Any:
    """
    Access a nested object in root by item sequence.

    Usage::
       data = {"nested": {"a_list": [{"finally": "target_data"}]}}
       value = dict_get(["nested", "a_list", 0, "finally"], "Not_found")

    """
    try:
        value = reduce(operator.getitem, items, root)
    except (KeyError, IndexError, TypeError) as e:
        logger.debug(e)
        return default
    return value


def match_string_link(
    text_item: str, data: str, need_key: str, matching_link_confs: List[Dict], render_context: Dict[str, Any]
) -> Any:
    try:
        link_name = None
        link_url = None
        link_conf = matching_link_confs[0]  # We only handle the first matching string_link
        match = link_conf["regex_compiled"].search(data)
        if match:
            render_content = match.groupdict()
            link_url = link_conf["url_template"].render(**render_content, **render_context)
            link_name = link_conf["name_template"].render(**render_content, **render_context)
        if link_name:
            ref_item = nodes.reference(link_name, link_name, refuri=link_url)
        else:
            # if no string_link match was made, we handle it as normal string value
            ref_item = nodes.Text(text_item)

    except Exception as e:
        logger.warning(
            f'Problems dealing with string to link transformation for value "{data}" of '
            f'option "{need_key}". Error: {e}'
        )
    else:
        return ref_item


def match_variants(option_value: Union[str, List], keywords: Dict, needs_variants: Dict) -> Union[str, List, None]:
    """
    Function to handle variant option management.

    :param option_value: Value assigned to an option
    :type option_value: Union[str, List]
    :param keywords: Data to use as filtering context
    :type keywords: Dict
    :param needs_variants: Needs variants data set in users conf.py
    :type needs_variants: Dict
    :return: A string, list, or None to be used as value for option.
    :rtype: Union[str, List, None]
    """

    def variant_handling(
        variant_definitions: List, variant_data: Dict, variant_pattern: Pattern
    ) -> Union[str, List, None]:
        filter_context = variant_data
        # filter_result = []
        no_variants_in_option = False
        variants_in_option = False
        for variant_definition in variant_definitions:
            # Test if definition is a variant definition
            check_definition = variant_pattern.search(variant_definition)
            if check_definition:
                variants_in_option = True
                # Separate variant definition from value to use for the option
                filter_string, output, _ = re.split(r"(:[\w':.\-\" ]+)$", variant_definition)
                filter_string = re.sub(r"^\[|[:\]]$", "", filter_string)
                filter_string = needs_variants[filter_string] if filter_string in needs_variants else filter_string
                try:
                    # https://docs.python.org/3/library/functions.html?highlight=compile#compile
                    filter_compiled = compile(filter_string, "<string>", "eval")
                    # Set filter_context as globals and not only locals in eval()!
                    # Otherwise, the vars not be accessed in list comprehensions.
                    if filter_compiled:
                        eval_result = bool(eval(filter_compiled, filter_context))
                    else:
                        eval_result = bool(eval(filter_string, filter_context))
                    # First matching variant definition defines the output
                    if eval_result:
                        no_variants_in_option = False
                        return output.lstrip(":")
                except Exception as e:
                    logger.warning(f'There was an error in the filter statement: "{filter_string}". ' f"Error Msg: {e}")
            else:
                no_variants_in_option = True

        if no_variants_in_option and not variants_in_option:
            return None

        # If no variant-rule is True, set to last, variant-free option. If this does not exist, set to None.
        defaults_to = variant_definitions[-1]
        if variants_in_option and variant_pattern.search(defaults_to):
            return None
        return re.sub(r"[;,] ", "", defaults_to)

    split_pattern = r"([\[\]]{1}[\w=:'. \-\"]+[\[\(\{]{1}[\w=,.': \-\"]*[\]\)\}]{1}[\[\]]{1}:[\w.\- ]+)|([\[\]]{1}[\w=:.'\-\[\] \"]+[\[\]]{1}:[\w.\- ]+)|([\w.: ]+[,;]{1})"
    variant_rule_pattern = r"^[\w'=,:.\-\"\[\] ]+:[\w'=:.\-\"\[\] ]+$"
    variant_splitting = re.compile(split_pattern)
    variant_rule_matching = re.compile(variant_rule_pattern)

    # Handling multiple variant definitions
    if isinstance(option_value, str):
        multiple_variants: List = variant_splitting.split(rf"""{option_value}""")
        multiple_variants: List = [
            re.sub(r"^([;, ]+)|([;, ]+$)", "", i) for i in multiple_variants if i not in (None, ";", "", " ")
        ]
        if len(multiple_variants) == 1 and not variant_rule_matching.search(multiple_variants[0]):
            return option_value
        new_option_value = variant_handling(multiple_variants, keywords, variant_rule_matching)
        if new_option_value is None:
            return option_value
        return new_option_value
    elif isinstance(option_value, (list, set, tuple)):
        multiple_variants: List = list(option_value)
        # In case an option value is a list (:tags: open; close), and does not contain any variant definition,
        # then return the unmodified value
        options = all([bool(not variant_rule_matching.search(i)) for i in multiple_variants])
        if options:
            return option_value
        new_option_value = variant_handling(multiple_variants, keywords, variant_rule_matching)
        return new_option_value
    else:
        return option_value


T = TypeVar("T")


def unwrap(obj: Optional[T]) -> T:
    assert obj is not None
    return obj


def node_match(node_types):
    """
    Returns a condition function for doctuils.nodes.findall()

    It takes a single or a list of node-types, if a findall() finds that node-type, the node
    get returned by findall() inside a generator-object.

    Use it like::

    for node_need in doctree.findall(node_mathc([Need, NeedTable])):
        if isinstance(node_nee, Need):
            pass  # some need voodoo
        elif isinstance(node_nee, NeedTable):
            pass  # some needtable voodoo
        else:
            raise Exception('Not requested node type')

    :param node_types: List of docutils node types
    :return: function, which can be used as constraint-function for docutils findall()
    """
    if not isinstance(node_types, list):
        node_types = [node_types]

    def condition(node, node_types=node_types):
        return any(isinstance(node, x) for x in node_types)

    return condition


def add_doc(env: BuildEnvironment, docname: str):
    """Stores a docname, to know later all need-relevant docs"""
    if docname not in env.needs_all_docs:
        env.needs_all_docs.append(docname)
