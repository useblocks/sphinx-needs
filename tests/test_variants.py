from pathlib import Path

import pytest
from sphinx.util.console import strip_colors

from sphinx_needs.utils import match_variants


@pytest.mark.parametrize(
    "option,context,variants,expected",
    [
        ("", {}, {}, None),
        ("a", {}, {}, None),
        ("a:yes", {}, {}, None),
        ("a:yes", {"a": False}, {}, None),
        ("a:yes", {"a": True}, {}, "yes"),
        ("a:yes, no", {"a": False}, {}, "no"),
        ("a:yes; no", {"a": False}, {}, "no"),
        ("[a and b]:yes, no", {"a": True, "b": True}, {}, "yes"),
        ("a:yes, no", {}, {"a": "True"}, "yes"),
        ("a:yes, no", {"b": 1}, {"a": "b == 1"}, "yes"),
        ("a:yes, no", {"b": 2}, {"a": "b == 1"}, "no"),
    ],
)
def test_match_variants(option, context, variants, expected):
    assert match_variants(option, context, variants) == expected


@pytest.mark.parametrize(
    "test_app",
    [{"buildername": "html", "srcdir": "doc_test/variant_doc", "tags": ["tag_a"]}],
    indirect=True,
)
def test_variant_options_html(test_app):
    app = test_app
    app.build()

    warnings = strip_colors(app._warning.getvalue()).splitlines()
    assert warnings == [
        f"{Path(str(app.srcdir)) / 'index.rst'}:25: WARNING: Error in filter 'tag_c': name 'tag_c' is not defined [needs.variant]"
    ]

    html = Path(app.outdir, "index.html").read_text()
    assert "Tags Example" in html
    assert "tags_implemented" in html
    assert "VA_003" in html

    assert "No ID" in html
    assert "progress" in html
    assert "extension" in html

    assert "Test story" in html
    assert "ST_001" in html
    assert "Daniel Woste" in html
    assert "Randy Duodu" not in html

    assert "Custom Variant" in html
    assert "CV_0002" in html
    assert "open" in html
    assert "start" in html
    assert "commence" in html

    # Check if referenced link exists in html
    assert (
        '<div class="line">links outgoing: <span class="links"><span><a class="reference '
        'internal" href="#VA_003" title="SPEC_003">VA_003</a></span></span></div>'
        in html
    )


@pytest.mark.parametrize(
    "test_app",
    [{"buildername": "html", "srcdir": "doc_test/variant_options", "tags": ["tag_a"]}],
    indirect=True,
)
def test_empty_variant_options_html(test_app):
    app = test_app
    app.build()
    html = Path(app.outdir, "index.html").read_text()
    assert "Empty Variant Options" in html
    assert "EMPTY_VAR_003" in html
    assert (
        '<div class="line"><span class="needs_status"><span class="needs_label">status: </span><span '
        'class="needs_data">[tag_a]:open, unknown</span></span></div>' in html
    )

    assert (
        '<div class="line">links outgoing: <span class="links"><span><span class="needs_dead_link '
        'forbidden">[tag_c]:CV_0002</span>, <span class="needs_dead_link forbidden">[tag_b]:VA_003</span>, '
        '<span class="needs_dead_link '
        'forbidden">unknown</span></span></span></div>' in html
    )
