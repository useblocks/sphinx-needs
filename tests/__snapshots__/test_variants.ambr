# serializer version: 1
# name: test_empty_variant_options_html[test_app0]
  dict({
    'current_version': '',
    'versions': dict({
      '': dict({
        'needs': dict({
          'EMPTY_VAR_003': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'EMPTY_VAR_003',
            'lineno': 4,
            'section_name': 'Empty Variant Options Handling Test',
            'sections': list([
              'Empty Variant Options Handling Test',
            ]),
            'status': '[tag_a]:open, unknown',
            'title': 'Empty Variant Options',
            'type': 'spec',
            'type_name': 'Specification',
          }),
        }),
        'needs_amount': 1,
        'needs_defaults_removed': True,
      }),
    }),
  })
# ---
# name: test_variant_options_html[test_app0]
  dict({
    'current_version': '',
    'versions': dict({
      '': dict({
        'needs': dict({
          'CV_0002': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'CV_0002',
            'lineno': 18,
            'section_name': 'Variant Handling Test',
            'sections': list([
              'Variant Handling Test',
            ]),
            'status': 'open',
            'tags': list([
              'commence',
              'start',
              'begin',
            ]),
            'title': 'Custom Variant',
            'type': 'spec',
            'type_name': 'Specification',
            'value': 'start',
          }),
          'SPEC_003': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'SPEC_003',
            'lineno': 25,
            'section_name': 'Variant Handling Test',
            'sections': list([
              'Variant Handling Test',
            ]),
            'status': 'open',
            'title': 'Variant Specification',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'SPEC_004': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'SPEC_004',
            'lineno': 29,
            'section_name': 'Variant Handling Test',
            'sections': list([
              'Variant Handling Test',
            ]),
            'status': 'unknown',
            'title': 'Unknown Variant',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'SP_38823': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'SP_38823',
            'lineno': 4,
            'section_name': 'Variant Handling Test',
            'sections': list([
              'Variant Handling Test',
            ]),
            'status': 'progress',
            'tags': list([
              'school',
              'extension',
              'needs',
            ]),
            'title': 'No ID',
            'type': 'spec',
            'type_name': 'Specification',
          }),
          'ST_001': dict({
            'author': 'Daniel Woste',
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'ST_001',
            'lineno': 12,
            'section_name': 'Variant Handling Test',
            'sections': list([
              'Variant Handling Test',
            ]),
            'status': 'close',
            'title': 'Test story',
            'type': 'story',
            'type_name': 'User Story',
          }),
          'VA_003': dict({
            'docname': 'index',
            'external_css': 'external_link',
            'id': 'VA_003',
            'lineno': 8,
            'section_name': 'Variant Handling Test',
            'sections': list([
              'Variant Handling Test',
            ]),
            'status': 'tags_implemented',
            'title': 'Tags Example',
            'type': 'spec',
            'type_name': 'Specification',
          }),
        }),
        'needs_amount': 6,
        'needs_defaults_removed': True,
      }),
    }),
  })
# ---
